### PR TITLE
Add `normalize' method to blocks

### LINKED ISSUE #9738: "normalize" mechanism for StreamField blocks
### Is your proposal related to a problem?

When writing to a StreamField block, e.g. `page.body = [('heading', "Hello world!")]`, or setting a `default` value, we expect the block value to be passed in the block's native data type. This data type is often undocumented and non-obvious:

* for RichTextBlock, it's a `wagtail.rich_text.RichText` object, not a string
* for EmbedBlock, it's an EmbedValue object
* for StreamBlock, it's a `StreamValue` instance. (The list-of-tuples representation is only recognised by `StreamField.to_python`, not StreamBlock, which means it's only valid for the top level of the stream, not nested StreamBlocks.)
* For ListBlock, it used to be a Python list, but is now a `ListValue` instance as of Wagtail 4.1.

StreamValue (and possibly ListValue too, now) is especially troublesome, as the constructor needs to be passed a reference to the StreamBlock, which is hard to get hold of, and impossible in the case of setting a `default` (because it would have to be a reference to the StreamBlock that you're still in the middle of defining).

In these situations, we should follow Postel's Law and accept "close enough" data types whenever a block value is being supplied from user code: a string for RichTextBlock, a list-of-tuples for StreamBlock and so on.

### Describe the solution you'd like

The `Block` class should implement a `normalize(value)` method that accepts an input value in a range of allowable types, and returns that value in the block's native type. This differs from existing methods such as `to_python` in that it's not aiming to convert data between representations - just standardising data that "should" be in the native type. We would then run any user-supplied data through `normalize` before using it in the block's internal logic.

The default implementation in `Block` would simply return `value` unchanged, while blocks that act as containers for child blocks will recursively call `normalize` on the child values along with doing whatever conversion they require for themselves.

### Additional context

If we were to move the "sniffing various types to decide what to do" logic from StreamField.to_python to StreamBlock.normalize, I believe this would remove the remaining blocker to supporting other kinds of block besides StreamBlock as the top-level block of a StreamField, providing a solution to #2048.

### PR DESCRIPTION
**Pull request by jams2**
Closes #9738

### CODE REVIEW

**Comment by lb-** [comment]
Don't want to bikeshed naming - but should we use `normalise` (en-GB spelling). 

Not sure either way but we do try to use UK spelling for documentation - not sure on APIs though. Also, if there is a Django /Python convention for this kind of method we should just use that spelling (which is probably US spelling). 


Alternatively - something like `get_native_value` could be explored.

**Comment by gasman** [comment]
> Don't want to bikeshed naming - but should we use `normalise` (en-GB spelling).

I was slightly on the fence about that when writing #9738... I figured that it would be a similar case to 'color', where people are used to seeing the American spelling in code to the point that having to remember it as a special case for Wagtail would become annoying.

**Comment by lb-** [comment]
@jams2 are you up for rebasing this PR and getting it ready for another review (marking as non-draft)?

As noted above, `normalize` is fine.

**Comment by jams2** [comment]
Hi @lb- , yep, planning to get it done over the next few weeks.

File: wagtail/blocks/stream_block.py
Comment (jams2) [inline wagtail/blocks/stream_block.py:313]: This was copied more-or-less verbatim from `StreamField._to_python`. It probably requires some cooperation with `StreamValue` to honour the `normalize` contract in the lazy case - as is, if the input is the "list of dicts" representation, the subvalues won't be normalized.

File: wagtail/blocks/list_block.py
Comment (jams2) [inline wagtail/blocks/list_block.py:232]: It might be nice to make this (and `StreamBlock.normalize`) generic over iterable types, rather than requiring a `list`?

File: wagtail/tests/test_blocks.py
Comment (jams2) [inline wagtail/tests/test_blocks.py:4424]: This is required by the check in `ListBlock._item_is_in_block_format`. Could this be relaxed, like `StreamBlock`, which handles the absence of an ID?

Comment (jams2) [reply wagtail/tests/test_blocks.py:4424]: Move the `ListBlock` cases to the `ListBlock` tests.

**Comment by lb-** [comment]
Looks like this has been rebased. Might be worth a fresh review from either @gasman or @laymonage

**Comment by lb-** [comment]
Adding to 6.1 milestone, I think this is a long desired improvement and will also enable more headless support in Wagtail.

File: wagtail/blocks/list_block.py
Comment (gasman) [inline wagtail/blocks/list_block.py:247]: This will discard any existing block IDs. However, rather than try to replicate the `to_python` logic to address this, I think we can reasonably leave this case out. We don't want `normalize` to be _too_ powerful, lest developers end up just throwing all data representations at it instead of using the proper conversion method - and I don't think there's any practical reason for people to be using the JSON-ish representation except when explicitly getting data in and out of the database.

We probably do still want to support `page.body = "some json string"`, but I think that's best handled at the top-level StreamField.to_python after all, and should be moved back there.

File: wagtail/exceptions.py
Comment (gasman) [inline wagtail/exceptions.py:12]: I'd question whether we really need a dedicated exception class for this - it occurs when people use an incompatible type when passing a value to a block, and to me that sounds like a job for a plain TypeError. Having a subclass is effectively just saying "a TypeError occurred during the normalization step", and that's leaking an unnecessary implementation detail to the user - for most purposes developers don't need to know that there's a normalization step.

**Review by gasman — CHANGES_REQUESTED** [review]
One addition I think we should have here: the base definition of `Block.get_default` should become `self.normalize(self.meta.default)`. That should allow us to simplify or eliminate the custom logic in StructBlock.get_default and StreamBlock.get_default, as well as supporting more complex default values that require recursive normalisation.

File: wagtail/blocks/list_block.py
Comment (gasman) [reply wagtail/blocks/list_block.py:247]: Digging a bit deeper into this: initially it appeared that `StreamField.to_python` was called when unpacking a value from the database, and [`Creator.__set__`](https://github.com/wagtail/wagtail/blob/e653a4afbb7f61e0e27e74e60a96981ca1d27f60/wagtail/fields.py#L81) was the path taken when assigning to the field from Python code - which would have allowed us to keep these conversions entirely separate. However, Django also [calls `Field.to_python` during `Field.clean`](https://github.com/django/django/blob/5.0.4/django/db/models/fields/__init__.py#L829), which means that `Field.to_python` has to accept Python-ish values as well as JSON-ish ones.

In the old version of this code, `Field.to_python` performed value sniffing to distinguish between these cases, so `StreamBlock.to_python` only has to handle JSON-ish values:

https://github.com/wagtail/wagtail/blob/e653a4afbb7f61e0e27e74e60a96981ca1d27f60/wagtail/fields.py#L167-L172

In the new code, this logic is moved to `StreamBlock.normalize`, which doesn't seem quite right as it means that `StreamBlock.to_python` is taken out of the picture. However, if we're aiming to eliminate all StreamBlock-specific logic from `StreamField.to_python`, it does make sense to offload this to the block class - probably to `StreamBlock.to_python`.

I'm heading towards the conclusion that Django is forcing us to make `Block.to_python` into a "handle any type we throw at it" method, and thus there's no benefit to having a distinct `normalize` method after all (since anything that might benefit from it could just as well call `to_python` instead). However, I'll need to think some more about whether that's an API-breaking change.

Comment (gasman) [reply wagtail/blocks/list_block.py:247]: OK, that won't work without breaking changes. Here's what hopefully will, though...

Definitions:

* JSONish value: a block value as it appears in the database JSON representation, e.g. `[{"type": "...", "id": "...", "value": ...}]` for a StreamBlock
* Pythonic value: any value for the block that might be passed from Python code, such as the tuple representation or StreamValue for StreamBlock, or a string or RichText for RichTextBlock
* Normalized pythonic value: the value in its "correct" type as seen when we read the value back from the StreamField, such as a StreamValue or RichText

In the current codebase, `Block.to_python` must accept a JSONish value and return a normalized pythonic value. (If it needs to recurse over children, it calls `to_python` on them.)

As of this PR, `Block.normalize` must accept a Pythonic value and return a normalized pythonic value. (If it needs to recurse over children, it calls `normalize` on them.)

Add a new requirement: for a block to be valid at the top level of a StreamField (currently only the case for StreamBlock), its `to_python` method must recognise when it has been passed a Pythonic value, and handle it equivalently to `normalize`. (Since `to_python`'s functionality is effectively a superset of `normalize`, it's likely that such a block would just implement a `to_python` method that accepts both Pythonic and JSONish values, and alias `normalize` to it.)

Now, `StreamField.to_python` can and `Creator.__set__` can simply delegate to the top-level block's `.to_python` method.

File: wagtail/blocks/stream_block.py
Comment (gasman) [inline wagtail/blocks/stream_block.py:326]: Worth noting that now StreamField always uses JSONField as the underlying database field, there is no circumstance where StreamField will be passed a string - JSON or otherwise - other than user code doing that explicitly with `page.body = "..."`. Indeed, the stated use-case for `raw_text` is no longer possible because converting a text field (containing non-JSON text) to StreamField will fail with a check constraint error. I think it would make sense to deprecate this.

File: wagtail/exceptions.py
Comment (jams2) [reply wagtail/exceptions.py]: Ok, I'd put it in thinking it _might_ be useful to distinguish between "this is an unexpected type error" and "normalize was called with the wrong arguments" at recursive call sites, but in practice those are the same thing or it's not really needed. Removed in 6b6550ba06

File: wagtail/blocks/list_block.py
Comment (jams2) [reply wagtail/blocks/list_block.py]: If `normalize` is implemented as an alias of `to_python`, is there any value to keeping `normalize` at all? It looks like most of the machinery for recursively calling `to_python` on subvalues is in place, the exception being `StreamBlock` which defers to `StreamValue`, which may or may not call `to_python` on subvalues, depending on the `is_lazy` argument.

Comment (gasman) [reply wagtail/blocks/list_block.py]: I think we do need to keep `normalize`, to account for existing block classes in non-Wagtail code that have a `to_python` method that only accepts JSONish values. (These classes won't implement `normalize`, so they'll just use the base `normalize` that returns the value unchanged.) If we don't have a separate `normalize` endpoint, then a container block handling a pythonic value and recursing over children will end up calling the child's `to_python` with a pythonic value, which isn't correct.

(A heads-up that I'm also working on this bit right now, so you might want to hold off to save duplicating work :-) )

File: wagtail/blocks/stream_block.py
Comment (gasman) [reply wagtail/blocks/stream_block.py]: Scratch that - it does happen when unpacking revision records (at least, ones created prior to the migration to JSONField).

File: wagtail/blocks/list_block.py
Comment (gasman) [reply wagtail/blocks/list_block.py]: Done in https://github.com/gasman/wagtail/commit/e1ec5a3caf964035d137a3bd3be3e9811cd92129

**Comment by gasman** [comment]
Superseded by #11850

**Pull request closed without merging**
